src/app/codegen-interaction-model/Model.cpp

```diff
@@ -0,0 +1,267 @@
+    }
+
+    // TODO: set entry flags:
+    //   info->flags.Set(ClusterQualityFlags::kDiagnosticsData)
+}
+
+/// Converts a EmberAfCluster into a ClusterEntry
+InteractionModel::ClusterEntry ClusterEntryFrom(EndpointId endpointId, const EmberAfCluster & cluster)
+{
+    InteractionModel::ClusterEntry entry;
+
+    entry.path = ConcreteClusterPath(endpointId, cluster.clusterId);
+    LoadClusterInfo(entry.path, cluster, &entry.info);
+
+    return entry;
+}
+
+/// Finds the first server cluster entry for the given endpoint data starting at [start_index]
+///
+/// Returns an invalid entry if no more server clusters are found
+InteractionModel::ClusterEntry FirstServerClusterEntry(EndpointId endpointId, const EmberAfEndpointType * endpoint,
+                                                       uint16_t start_index)
+{
+    for (unsigned i = start_index; i < endpoint->clusterCount; i++)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[i];
+        if (!IsServerMask(cluster.mask & CLUSTER_MASK_SERVER))
+        {
+            continue;
+        }
+
+        return ClusterEntryFrom(endpointId, cluster);
+    }
+
+    return InteractionModel::ClusterEntry::Invalid();
+}
+
+/// Load the cluster information into the specified destination
+void LoadAttributeInfo(const ConcreteAttributePath & path, const EmberAfAttributeMetadata & attribute,
+                       InteractionModel::AttributeInfo * info)
+{
+    if (attribute.attributeType == ZCL_ARRAY_ATTRIBUTE_TYPE)
+    {
+        info->flags.Set(InteractionModel::AttributeQualityFlags::kListAttribute);
+    }
+
+    // TODO: Set additional flags:
+    // info->flags.Set(InteractionModel::AttributeQualityFlags::kChangesOmitted)
+}
+
+InteractionModel::AttributeEntry AttributeEntryFrom(const ConcreteClusterPath & clusterPath,
+                                                    const EmberAfAttributeMetadata & attribute)
+{
+    InteractionModel::AttributeEntry entry;
+
+    entry.path = ConcreteAttributePath(clusterPath.mEndpointId, clusterPath.mClusterId, attribute.attributeId);
+    LoadAttributeInfo(entry.path, attribute, &entry.info);
+
+    return entry;
+}
+
+} // namespace
+
+CHIP_ERROR Model::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, InteractionModel::ReadState & state,
+                                AttributeValueEncoder & encoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR Model::WriteAttribute(const InteractionModel::WriteAttributeRequest & request, AttributeValueDecoder & decoder)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR Model::Invoke(const InteractionModel::InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
+                         InteractionModel::InvokeReply & reply)
+{
+    // TODO: this needs an implementation
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+EndpointId Model::FirstEndpoint()
```

Comment:
Suggest renaming `Model` to `CodeGenModel` or `EmberDataModel`. Namespacing-only makes it harder to search for given files or classes since you need a full-compiler-index-aware search to find the correct `Model` among the possibly many namespaced variants.

Reply:
Renamed.